rllib/utils/replay_buffers/reservoir_buffer.py

```diff
@@ -0,0 +1,109 @@
+from typing import Any, Dict
+import random
+
+# Import ray before psutil will make sure we use psutil's bundled version
+import ray  # noqa F401
+import psutil  # noqa E402
+
+from ray.rllib.utils.annotations import ExperimentalAPI, override
+from ray.rllib.utils.replay_buffers.replay_buffer import ReplayBuffer
+from ray.rllib.utils.typing import SampleBatchType
+from ray.rllib.execution.buffers.replay_buffer import warn_replay_capacity
+
+
+@ExperimentalAPI
+class ReservoirBuffer(ReplayBuffer):
+    """This buffer implements reservoir sampling.
+
+    The algorithm has been described by Jeffrey S. Vitter in "Random sampling
+    with a reservoir". See https://www.cs.umd.edu/~samir/498/vitter.pdf for
+    the full paper.
+    """
+
+    def __init__(self, capacity: int = 10000, storage_unit: str = "timesteps"):
+        """Initializes a ReservoirBuffer instance.
+
+        Args:
+            capacity: Max number of timesteps to store in the FIFO
+                buffer. After reaching this number, older samples will be
+                dropped to make space for new ones.
+            storage_unit: Either 'sequences' or 'timesteps'. Specifies how
+                experiences are stored.
+        """
+        ReplayBuffer.__init__(self, capacity, storage_unit)
+        self._num_add_calls = 0
+        self._num_evicted = 0
+
+    @ExperimentalAPI
+    @override(ReplayBuffer)
+    def add(self, batch: SampleBatchType, **kwargs) -> None:
+        """Adds a batch of experiences.
+
+        Args:
+            batch: SampleBatch to add to this buffer's storage.
+        """
+        # Update add counts.
+        self._num_add_calls += 1
+        # Update our timesteps counts.
+        self._num_timesteps_added += batch.count
+        self._num_timesteps_added_wrap += batch.count
+
+        if self._num_timesteps_added < self.capacity:
+            ReplayBuffer.add(self, batch)
+        else:
+            # Eviction of older samples has already started (buffer is "full")
+            self._eviction_started = True
+            idx = random.randint(0, self._num_add_calls - 1)
```

Comment:
mostly for my own curiosity, can you add a reference to the algorithm somehow?
I am a bit confused about randomly sampling from [0, num_add_calls].
over time, vast majority of the samples collected will get dropped with this approach?

Reply:
The algorithm itself is referenced in the class docstring.
Yes, a vast majority of samples will get dropped.
This is just me following the paper step by step.
Is far as I understand reservoir sampling, this holds true any implementation of reservoir buffers - if you want to sample all with equal probability from all seen samples any time, you would need infinite memory to hold all the samples. If you want to have a chance to sample your first experience after collecting a multiple of your buffer size in experiences, there needs to be a probability for new samples to get dropped.

I am curious and very open to discussion and a meeting about this.